File: src/ros2_bridge/mappers/geometry_msgs/wrench.rs
//! Typed mapper for `geometry_msgs/msg/Wrench`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn wrench_to_bus(msg: ros_env::geometry_msgs::msg::Wrench) -> crate::geometry_msgs::msg::v1::Wrench {
    crate::geometry_msgs::msg::v1::Wrench {
        force: Some(crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_bus(msg.force)),
        torque: Some(crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_bus(msg.torque)),
    }
}

pub(crate) fn wrench_to_ros(bus: crate::geometry_msgs::msg::v1::Wrench) -> ros_env::geometry_msgs::msg::Wrench {
    ros_env::geometry_msgs::msg::Wrench {
        force: crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_ros(bus.force.unwrap_or_default()),
        torque: crate::ros2_bridge::mappers::geometry_msgs::vector3::vector3_to_ros(bus.torque.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct GeometryMsgsWrenchMapper;

impl TypedTopicMapper for GeometryMsgsWrenchMapper {
    type Ros = ros_env::geometry_msgs::msg::Wrench;
    type Bus = crate::geometry_msgs::msg::v1::Wrench;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(wrench_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(wrench_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/geometry_msgs/wrench.py
"""Generated mapper for `geometry_msgs/msg/Wrench`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.geometry_msgs.vector3 import vector3_to_bus, vector3_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.geometry_msgs.msg.v1 import Wrench as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def wrench_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.force.CopyFrom(vector3_to_bus(msg.force))
    bus.torque.CopyFrom(vector3_to_bus(msg.torque))
    return bus


def wrench_to_ros(bus):
    from geometry_msgs.msg import Wrench as RosMsg

    out = RosMsg()
    out.force = vector3_to_ros(bus.force)
    out.torque = vector3_to_ros(bus.torque)
    return out


class GeometryMsgsWrenchMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from geometry_msgs.msg import Wrench as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return wrench_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return wrench_to_ros(bus)
